contrib/IECoreUSD/src/IECoreUSD/USDScene.cpp

```diff
@@ -0,0 +1,832 @@
+	}
+}
+
+IECore::PointsPrimitivePtr convert( pxr::UsdGeomPoints points, pxr::UsdTimeCode time )
+{
+	pxr::UsdAttribute attr = points.GetPointsAttr();
+
+	pxr::VtVec3fArray pointsData;
+
+	attr.Get( &pointsData, time );
+
+	IECore::V3fVectorDataPtr positionData = convert( pointsData );
+
+	IECore::PointsPrimitivePtr newPoints = new IECore::PointsPrimitive( positionData );
+
+	convertPrimVars( points, time, newPoints );
+	return newPoints;
+}
+
+IECore::CurvesPrimitivePtr convert( pxr::UsdGeomCurves curves, pxr::UsdTimeCode time )
+{
+	pxr::UsdAttribute vertexCountsAttr = curves.GetCurveVertexCountsAttr();
+	pxr::VtIntArray vertexCountsData;
+	vertexCountsAttr.Get( &vertexCountsData, time );
+	IECore::IntVectorDataPtr countData = convert( vertexCountsData );
+
+	pxr::UsdAttribute attr = curves.GetPointsAttr();
+	pxr::VtVec3fArray pointsData;
+	attr.Get( &pointsData, time );
+	IECore::V3fVectorDataPtr positionData = convert( pointsData );
+
+	IECore::CurvesPrimitivePtr newCurves = new IECore::CurvesPrimitive( countData, IECore::CubicBasisf::linear(), false, positionData );
+
+	convertPrimVars( curves, time, newCurves );
+	return newCurves;
+}
+
+IECore::MeshPrimitivePtr convert( pxr::UsdGeomMesh mesh, pxr::UsdTimeCode time )
+{
+	pxr::UsdAttribute subdivSchemeAttr = mesh.GetSubdivisionSchemeAttr();
+
+	pxr::TfToken subdivScheme;
+	subdivSchemeAttr.Get( &subdivScheme );
+
+	pxr::UsdAttribute faceVertexCountsAttr = mesh.GetFaceVertexCountsAttr();
+
+	pxr::VtIntArray faceVertexCounts;
+	faceVertexCountsAttr.Get( &faceVertexCounts );
+	IECore::IntVectorDataPtr vertexCountData = convert( faceVertexCounts );
+
+	pxr::UsdAttribute faceVertexIndexAttr = mesh.GetFaceVertexIndicesAttr();
+	pxr::VtIntArray faceVertexIndices;
+	faceVertexIndexAttr.Get( &faceVertexIndices );
+
+	IECore::IntVectorDataPtr vertexIndicesData = convert( faceVertexIndices );
+
+	IECore::MeshPrimitivePtr newMesh = new IECore::MeshPrimitive( vertexCountData, vertexIndicesData );
+
+	pxr::UsdAttribute attr = mesh.GetPointsAttr();
+	pxr::VtVec3fArray pointsData;
+
+	attr.Get( &pointsData, time );
+
+	IECore::V3fVectorDataPtr positionData = convert( pointsData );
+
+	convertPrimVars( mesh, time, newMesh );
+	newMesh->variables["P"] = IECore::PrimitiveVariable( IECore::PrimitiveVariable::Vertex, positionData );
+
+	if( subdivScheme == pxr::UsdGeomTokens->catmullClark )
+	{
+		newMesh->setInterpolation( "catmullClark" );
+	}
+
+	return newMesh;
+}
+
+bool isConvertable( pxr::UsdPrim prim )
+{
+	pxr::UsdGeomMesh mesh = pxr::UsdGeomMesh::Get( prim.GetStage(), prim.GetPath() );
+	if( mesh )
+	{
+		return true;
+	}
+	pxr::UsdGeomPoints points = pxr::UsdGeomPoints::Get( prim.GetStage(), prim.GetPath() );
+	if( points )
+	{
+		return true;
+	}
+
+	pxr::UsdGeomCurves curves = pxr::UsdGeomCurves::Get( prim.GetStage(), prim.GetPath() );
+	if( curves )
+	{
+		return true;
+	}
+	return false;
+}
+
+IECore::ConstObjectPtr convert( pxr::UsdPrim prim, pxr::UsdTimeCode time )
```

Comment:
We tend to use registries for conversion functions so that we can add converters non-intrusively rather than having to manage a nasty central `if this then that else if this then...` thing. This approach would also mean that `isConvertible()` didn't have to duplicate exactly the same functionality. Maybe we don't need to do that right now, but it's worth it in the long run I think.

By the way, is it possible to ask USD "what type is this?", or do we literally have to say "is it A? is it B? is it C?" until we get a positive answer? Alembic is the latter, and it sucks, but we've still managed to build an extensible registry of conversions around it.

Reply:
> We tend to use registries for conversion functions so that we can add converters non-intrusively rather than having to manage a nasty central `if this then that else if this then...` thing. This approach would also mean that `isConvertible()` didn't have to duplicate exactly the same functionality. Maybe we don't need to do that right now, but it's worth it in the long run I think.

Understood. I'll work on this directly after this PR is merged

> By the way, is it possible to ask USD "what type is this?", or do we literally have to say "is it A? is it B? is it C?" until we get a positive answer? Alembic is the latter, and it sucks, but we've still managed to build an extensible registry of conversions around it.

I don't know and will investigate